### PR TITLE
is0502: check SDP clocks match IS-04 Sources

### PR DESCRIPTION
**Pull request by andrewbonney**
This adds a further SDP check to ensure that ts-refclk and IS-04 Source clocks match up.

### CODE REVIEW

File: nmostesting/suites/IS0502Test.py [hunk 1]
```diff
@@ -952,6 +952,81 @@ def test_16(self, test):
...
+
+            found_refclk = False
+            for sdp_line in is05_transport_file.split("\n"):
+                sdp_line = sdp_line.replace("\r", "")
+                ts_refclk = re.search(r"^a=ts-refclk:(.+)$", sdp_line)
+                if not ts_refclk:
+                    continue
+                found_refclk = True
+                if source["clock_name"] is None:
+                    return test.FAIL("SDP file includes ts-refclk but Source {} does not indicate a clock_name"
```
Comment (garethsb-sony) [inline nmostesting/suites/IS0502Test.py]: By the way, is having a non-`null` clock a requirement of a "periodic Source", like having a grain rate?

Comment (andrewbonney) [reply nmostesting/suites/IS0502Test.py]: I guess it depends. If the signal originated at the Sender then it probably ought to signal a clock (even if internal), but if the signal originated elsewhere and was just transformed by the Sender, the Sender's Node may not lock the signal to its own clock first.

**Comment by andrewbonney** [comment]
> Did this ought to allow a `traceable` PTP clock to be indicated as such in the SDP `ts-refclk` rather than with a GMID? I'm not sure... whereas in the SDP you can only indicate either/or, in the NMOS schema all props are required...

I'll take a look at this. I suspect a minor change may be required.

> Thinking about the ST 2110-10 `localmac` form. Can we check whether it has a MAC address that matches the Sender `interface_binding` -> Node `interface` `port_id`? Though what about ST 2002-7? Or if all Senders on the Node are sync'ed to the same internal clock, may the `localmac` form indicate a single `port_id` or use a common `chassis_id`?

I'm not sure we can guarantee that the relevant interface will always be signalled in interface_bindings, so I'd be inclined to avoid this check for now.

**Comment by garethsb-sony** [comment]
What do you mean by "relevant interface"? Why wouldn't it be one (of the ones) indicated in interface_bindings?

**Comment by andrewbonney** [comment]
There is no requirement in the spec to list every interface in `interface_bindings`. Whilst in our case we do, the 'default' interface (and so the one perhaps most likely to occur in a localmac line) is the control interface, which doesn't have to be listed. It's possible this may match the `chassis_id`, but it's also possible the `chassis_id` may not be represented as a MAC address.

**Comment by garethsb-sony** [comment]
Yes, sorry, I think we're at cross purposes. On the other side, ST 2110-10 says "Devices which are not referenced to IEEE1588-2008 shall signal using the extended form shown below, indicating _the MAC address of the sender_ using the token `localmac`". Isn't that the interface indicated in the sender's `interface_bindings`?

**Comment by andrewbonney** [comment]
Yes agreed, given that 2110-10 clearly defines the Sender as originating one RTP stream that seems clear it must match the one listed in the `interface_bindings`. I'll add that check.

**Comment by andrewbonney** [comment]
Added 'localmac' checking. There are a couple of assumptions here. First is that ts-refclk isn't defined globally in the SDP, and second is that the `interface_bindings` order matches the SDP ordering. I think these are relatively safe assumptions.

Unfortunately I don't have a straightforward way to test this fully in practice.

**Comment by garethsb-sony** [comment]
I'll have a go when I'm back in the UK.

**Comment by garethsb-sony** [comment]
> Added 'localmac' checking. There are a couple of assumptions here. First is that ts-refclk isn't defined globally in the SDP,

Although RFC 7273 permits session-level ts-refclk, ST 2110-10 Section 8.2 says "All stream descriptions shall have a media-level ts-refclk attribute as specified in IETF RFC 7273 section 4." Apparently this extra constraint on spec was meant to make things easier. 🤷‍♂ Anyway, I think it means your assumption is OK for ST 2110.

> and second is that the `interface_bindings` order matches the SDP ordering. I think these are relatively safe assumptions.

I think neither ST 2110-10 Section 8.3 nor RFC 7104 require any particular order of the media descriptions, or of the `a=group:DUP` id list. Maybe it could be possible to match source or destination IP addresses to the `transport_params` leg and then to `interface_bindings`, but I think the assumption is reasonable and probably the opposite ordering could be used as a torture case for clients. ;-)

I have one question as to whether ST 2022-7 stream sh/could use same (primary?) `localmac=` address on both legs, to indicate the two streams are synchronized, or not though...

**Comment by andrewbonney** [comment]
> I have one question as to whether ST 2022-7 stream sh/could use same (primary?) `localmac=` address on both legs, to indicate the two streams are synchronized, or not though...

I'm unsure. My interpretation from ST.2110-10's use of the term 'sender' was that streams originating from different interfaces would require use of those interfaces' MACs, but it's not 100% clear.

**Comment by garethsb-sony** [comment]
I'm wondering how this would work with a single media description using a `source-filter` to indicate [separate source addresses](https://github.com/AMWA-TV/nmos-device-connection-management/blob/v1.1.x/docs/4.1.%20Behaviour%20-%20RTP%20Transport%20Type.md#separate-source-addresses)?

**Comment by andrewbonney** [comment]
I'm not sure. I haven't seen any implementations using that mechanism in practice so far.

**Comment by garethsb-sony** [comment]
Summary of a Slack conversation:

* Using one media description with a source-filter to indicate multiple source addresses seems to be equivalent to using multiple media descriptions, one for each of the source addresses but the same destination address.
* ST 2022-7 doesn't appear to have any reference to SDP and ST 2110-10 says you can't have multiple media with the same source _and_ destination, but otherwise you're OK.
* If you want to use `localmac` reference clocks, you should therefore use one media description per leg, rather than the source-filter approach.

**Pull request merged**